persistence/src/main/java/com/epam/esm/persistence/repository/impl/GiftCertificateRepositoryImpl.java

```diff
@@ -1,120 +1,103 @@
+        return entityManager.find(GiftCertificate.class, id);
     }
 
     @Override
-    public GiftCertificate update(Long id, GiftCertificate entity) {
-        jdbcTemplate.update(UPDATE_QUERY,
-                entity.getName(),
-                entity.getDescription(),
-                entity.getPrice(),
-                entity.getDuration(),
-                id
-        );
-        return entity;
+    public GiftCertificate update(GiftCertificate entity) {
+        return entityManager.merge(entity);
     }
 
     @Override
+    @Transactional
     public boolean delete(Long id) {
-        return jdbcTemplate.update(DELETE_BY_ID_QUERY, id) == MIN_AFFECTED_ROWS;
+        GiftCertificate certificate = entityManager.find(GiftCertificate.class, id);
+        if (certificate == null) {
+            return false;
+        }
+        if (hasAssociatedOrders(certificate)) {
+            throw new RepositoryException(RepositoryErrorCode.DELETION_FORBIDDEN, GIFT_CERTIFICATE_ENTITY_NAME, id,
+                    ORDER_ENTITY_NAME);
+        }
+        entityManager.remove(certificate);
+        return true;
     }
 
-    @Override
-    public List<GiftCertificate> findWithFilters(QueryFiltersConfig config) {
-        PreparedStatementCreator statementCreator = new GiftCertificatePreparedStatementCreator(config);
-        return jdbcTemplate.query(statementCreator, certificateRowMapper);
+    /**
+     * Checks if gift certificate has orders associated with it.
+     *
+     * @param certificate Gift certificate entity
+     * @return {@code true} if gift certificate has orders associated with it
+     */
+    private boolean hasAssociatedOrders(GiftCertificate certificate) {
+        TypedQuery<Order> query = entityManager.createQuery(SELECT_ASSOCIATED_ORDERS, Order.class);
+        query.setParameter("giftCertificate", certificate);
+        List<Order> resultList = query.getResultList();
+        return !resultList.isEmpty();
     }
 
     @Override
-    public List<Tag> findAssociatedTags(Long certificateId) {
-        return jdbcTemplate.query(SELECT_ASSOCIATED_TAGS_QUERY, tagRowMapper, certificateId);
+    public List<GiftCertificate> findWithFilters(GiftCertificatesFilterConfig config, int page, int size) {
+        CriteriaQuery<GiftCertificate> criteriaQuery = config.asCriteriaQuery(entityManager.getCriteriaBuilder());
+        TypedQuery<GiftCertificate> typedQuery = entityManager.createQuery(criteriaQuery);
+        typedQuery.setFirstResult((page - 1) * size);
+        typedQuery.setMaxResults(size);
+        return typedQuery.getResultList();
     }
 
     @Override
-    public void addTagAssociation(Long certificateId, Long tagId) {
-        jdbcTemplate.update(ADD_TAG_ASSOCIATION_QUERY, certificateId, tagId);
+    public List<Tag> findAssociatedTags(Long certificateId) {
+        return new ArrayList<>(entityManager.find(GiftCertificate.class, certificateId).getAssociatedTags());
```

Comment:
do we need this method? Can we use findById  in certificate service?

Reply:
This method can be replaced with findById and getting tags from certificate entity in the service.